include/targets/generic/shared_memory_cache_helper.h

```diff
@@ -1,68 +1,282 @@
+     the cache which any other thread can access.  The data is stored
+     in a coalesced order with element size atom_t<T>.
+
+     The dimensions of the cache is determined by a call to
+     D::dims(target::block_dim()), and D defaults to having dimensions
+     equal to the block dimensions.
+
+     A byte offset into the shared memory region can be specified with
+     the type O, and is given by
+     O::shared_mem_size(target::block_dim()) if O is not void.
+   */
+  template <typename T, typename D = DimsBlock, typename O = void>
+  class SharedMemoryCache : SharedMemory<atom_t<T>, SizeDims<D, sizeof(T) / sizeof(atom_t<T>)>, O>
+  {
+    using Smem = SharedMemory<atom_t<T>, SizeDims<D, sizeof(T) / sizeof(atom_t<T>)>, O>;
+
+  public:
+    using value_type = T;
+    using dims_type = D;
+    using offset_type = O;
+    using Smem::shared_mem_size;
+
+  private:
+    const dim3 block;
+    const int stride;
+    using Smem::sharedMem;
+    using atom_t = atom_t<T>;
+    static_assert(sizeof(T) % 4 == 0, "Shared memory cache does not support sub-word size types");
+
+    // The number of elements of type atom_t that we break T into for optimal shared-memory access
+    static constexpr int n_element = sizeof(T) / sizeof(atom_t);
+
+    // used to avoid instantiation of load functions if unused, in case T is not a valid return type (e.g. C array)
+    template <typename dummy = void> using maybeT = std::conditional_t<std::is_same_v<dummy, void>, T, void>;
+
+    __device__ __host__ inline void save_detail(const T &a, int x, int y, int z) const
+    {
+      atom_t tmp[n_element];
+      memcpy(tmp, (void *)&a, sizeof(T));
+      int j = (z * block.y + y) * block.x + x;
+#pragma unroll
+      for (int i = 0; i < n_element; i++) sharedMem()[i * stride + j] = tmp[i];
+    }
+
+    template <typename dummy = void> __device__ __host__ inline maybeT<dummy> load_detail(int x, int y, int z) const
+    {
+      atom_t tmp[n_element];
+      int j = (z * block.y + y) * block.x + x;
+#pragma unroll
+      for (int i = 0; i < n_element; i++) tmp[i] = sharedMem()[i * stride + j];
+      T a;
+      memcpy((void *)&a, tmp, sizeof(T));
+      return a;
+    }
+
+    /**
+       @brief Dummy instantiation for the host compiler
+    */
+    template <bool is_device, typename dummy = void> struct sync_impl {
+      void operator()() { }
+    };
+
+    /**
+       @brief Synchronize the cache when on the device
+    */
+    template <typename dummy> struct sync_impl<true, dummy> {
+      __device__ inline void operator()() { __syncthreads(); }
+    };
+
+  public:
+    /**
+       @brief Constructor for SharedMemoryCache.
+    */
+    constexpr SharedMemoryCache() : block(D::dims(target::block_dim())), stride(block.x * block.y * block.z)
+    {
+      // sanity check
+      static_assert(shared_mem_size(dim3 {32, 16, 8})
+                    == Smem::get_offset(dim3 {32, 16, 8}) + SizeDims<D>::size(dim3 {32, 16, 8}) * sizeof(T));
```

Comment:
Are these sizes arbitrary?

Reply:
Yes, I just picked some arbitrary static sizes for the sanity check.